src/libraries/System.Private.CoreLib/src/System/IO/Strategies/AsyncWindowsFileStreamStrategy.ValueTaskSource.cs

```diff
@@ -3,47 +3,39 @@
 
 using System.Buffers;
 using System.Diagnostics;
-using System.Runtime.InteropServices;
 using System.Threading;
 using System.Threading.Tasks.Sources;
-using TaskSourceCodes = System.IO.Strategies.FileStreamHelpers.TaskSourceCodes;
 
 namespace System.IO.Strategies
 {
     internal sealed partial class AsyncWindowsFileStreamStrategy : WindowsFileStreamStrategy
     {
-        /// <summary>
-        /// Type that helps reduce allocations for FileStream.ReadAsync and FileStream.WriteAsync.
-        /// </summary>
-        private sealed unsafe class ValueTaskSource : IValueTaskSource<int>, IValueTaskSource
+        /// <summary>Reusable IValueTaskSource for FileStream ValueTask-returning async operations.</summary>
+        private sealed unsafe class ValueTaskSource : ManualResetEventSlim, IValueTaskSource<int>, IValueTaskSource
         {
+            // Base MRES is used to prevent the callback from proceeding until the async initiating has fully
+            // completed, e.g. we've finished registering for cancellation after kicking off the operation.
             internal static readonly IOCompletionCallback s_ioCallback = IOCallback;
+
             internal readonly PreAllocatedOverlapped _preallocatedOverlapped;
             private readonly AsyncWindowsFileStreamStrategy _strategy;
-            private MemoryHandle _handle;
+            internal MemoryHandle _memoryHandle;
             internal ManualResetValueTaskSourceCore<int> _source; // mutable struct; do not make this readonly
             private NativeOverlapped* _overlapped;
             private CancellationTokenRegistration _cancellationRegistration;
-            private long _result; // Using long since this needs to be used in Interlocked APIs
-#if DEBUG
-            private bool _cancellationHasBeenRegistered;
-#endif
 
             internal ValueTaskSource(AsyncWindowsFileStreamStrategy strategy)
             {
                 _strategy = strategy;
-                _preallocatedOverlapped = new PreAllocatedOverlapped(s_ioCallback, this, null);
-
                 _source.RunContinuationsAsynchronously = true;
+                _preallocatedOverlapped = new PreAllocatedOverlapped(s_ioCallback, this, null);
             }
 
-            internal NativeOverlapped* Configure(ReadOnlyMemory<byte> memory)
+            internal NativeOverlapped* PrepareForOperation(ReadOnlyMemory<byte> memory)
             {
-                _result = TaskSourceCodes.NoResult;
-
-                _handle = memory.Pin();
+                base.Reset();
```

Comment:
`AsyncWindowsFileStreamStrategy.TryToReuse` calls `_source.Reset`:

https://github.com/dotnet/runtime/blob/f3eafd7b3be3e869cc9d1f0bc0ec5e93c40e2f17/src/libraries/System.Private.CoreLib/src/System/IO/Strategies/AsyncWindowsFileStreamStrategy.cs#L106-L108

this one calls `base.Reset`. Could you please add comments that explain what these two resets do and why are they in two different places?

Reply:
They were different Reset methods.  One was on the ManualResetEventSlim, one was on the ManualResetValueTaskSourceCore.

But, I decided to revisit all of this, and changed the code to remove the MRES I'd added.  It's a middle ground between what was there and what I'd done, and avoids the blocking I'd added.